In [5]:
import sqlalchemy

In [6]:
sqlalchemy.create_engine("sqlite:///chinook.db")

Engine(sqlite:///chinook.db)

In [7]:
#%%capture
%load_ext sql
%sql sqlite:///chinook.db

'Connected: @chinook.db'

In [8]:
%%sql
SELECT * FROM invoice
LIMIT 10

 * sqlite:///chinook.db
Done.


[(1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84),
 (2, 30, '2017-01-03 00:00:00', '230 Elgin Street', 'Ottawa', 'ON', 'Canada', 'K2P 1L7', 9.9),
 (3, 40, '2017-01-05 00:00:00', '8, Rue Hanovre', 'Paris', 'None', 'France', '75002', 1.98),
 (4, 18, '2017-01-06 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 7.92),
 (5, 27, '2017-01-07 00:00:00', '1033 N Park Ave', 'Tucson', 'AZ', 'USA', '85719', 16.83),
 (6, 31, '2017-01-10 00:00:00', '194A Chain Lake Drive', 'Halifax', 'NS', 'Canada', 'B3S 1C5', 1.98),
 (7, 49, '2017-01-12 00:00:00', 'Ordynacka 10', 'Warsaw', 'None', 'Poland', '00-358', 10.89),
 (8, 59, '2017-01-13 00:00:00', '3,Raj Bhavan Road', 'Bangalore', 'None', 'India', '560001', 9.9),
 (9, 18, '2017-01-18 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 8.91),
 (10, 31, '2017-01-18 00:00:00', '194A Chain Lake Drive', 'Halifax', 'NS', 'Canada', 'B3S 1C5', 1.98)]

In [8]:
%%sql
SELECT * FROM customer
LIMIT 10

 * sqlite:///chinook.db
Done.


[(1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3),
 (2, 'Leonie', 'Köhler', None, 'Theodor-Heuss-Straße 34', 'Stuttgart', None, 'Germany', '70174', '+49 0711 2842222', None, 'leonekohler@surfeu.de', 5),
 (3, 'François', 'Tremblay', None, '1498 rue Bélanger', 'Montréal', 'QC', 'Canada', 'H2G 1A7', '+1 (514) 721-4711', None, 'ftremblay@gmail.com', 3),
 (4, 'Bjørn', 'Hansen', None, 'Ullevålsveien 14', 'Oslo', None, 'Norway', '0171', '+47 22 44 22 22', None, 'bjorn.hansen@yahoo.no', 4),
 (5, 'František', 'Wichterlová', 'JetBrains s.r.o.', 'Klanova 9/506', 'Prague', None, 'Czech Republic', '14700', '+420 2 4172 5555', '+420 2 4172 5555', 'frantisekw@jetbrains.com', 4),
 (6, 'Helena', 'Holý', None, 'Rilská 3174/6', 'Prague', None, 'Czech Republic', '14300', '+420 2 4177 0449', None, 'hholy@gmail.com', 5),
 (7, 'Astrid', 'Gruber', None, 'Rotenturmstraße 4, 1010 Innere Stadt', 'Vienne', None, 'Austria', '1010', '+43 01 5134505', None, 'astrid.gruber@apple.at', 5),
 (8, 'Daan', 'Peeters', None, 'Grétrystraat 63', 'Brussels', None, 'Belgium', '1000', '+32 02 219 03 03', None, 'daan_peeters@apple.be', 4),
 (9, 'Kara', 'Nielsen', None, 'Sønder Boulevard 51', 'Copenhagen', None, 'Denmark', '1720', '+453 3331 9991', None, 'kara.nielsen@jubii.dk', 4),
 (10, 'Eduardo', 'Martins', 'Woodstock Discos', 'Rua Dr. Falcão Filho, 155', 'São Paulo', 'SP', 'Brazil', '01007-010', '+55 (11) 3033-5446', '+55 (11) 3033-4564', 'eduardo@woodstock.com.br', 4)]

In [9]:
%%sql
SELECT
    name,
    type
FROM sqlite_master
WHERE type IN ("table","view");

 * sqlite:///chinook.db
Done.


[('album', 'table'),
 ('artist', 'table'),
 ('customer', 'table'),
 ('employee', 'table'),
 ('genre', 'table'),
 ('invoice', 'table'),
 ('invoice_line', 'table'),
 ('media_type', 'table'),
 ('playlist', 'table'),
 ('playlist_track', 'table'),
 ('track', 'table'),
 ('customer_sale', 'view')]

In [10]:
%%sql
SELECT 
    g.name genre_name,
    COUNT(t.track_id) number_of_track,
    COUNT(t.track_id) * 100.0 / SUM(COUNT(t.track_id)) over() percentage
FROM genre g
INNER JOIN track t ON t.genre_id=g.genre_id
INNER JOIN invoice_line il ON il.track_id=t.track_id
INNER JOIN invoice i ON i.invoice_id=il.invoice_id
WHERE i.billing_country='USA'
GROUP BY 1
ORDER BY 2 DESC;

 * sqlite:///chinook.db
Done.


[('Rock', 561, 53.37773549000951),
 ('Alternative & Punk', 130, 12.369172216936251),
 ('Metal', 124, 11.798287345385347),
 ('R&B/Soul', 53, 5.042816365366318),
 ('Blues', 36, 3.4253092293054235),
 ('Alternative', 35, 3.330161750713606),
 ('Latin', 22, 2.093244529019981),
 ('Pop', 22, 2.093244529019981),
 ('Hip Hop/Rap', 20, 1.9029495718363463),
 ('Jazz', 14, 1.3320647002854424),
 ('Easy Listening', 13, 1.236917221693625),
 ('Reggae', 6, 0.570884871550904),
 ('Electronica/Dance', 5, 0.47573739295908657),
 ('Classical', 4, 0.38058991436726924),
 ('Heavy Metal', 3, 0.285442435775452),
 ('Soundtrack', 2, 0.19029495718363462),
 ('TV Shows', 1, 0.09514747859181731)]

* The most popular genre is Rock which has be sold for 561 tracks and accounts for more than half of the total sale.
* According to the ranking list above, we recommend Red Tone whose music genre is Punk, Slim Jim Bites whose music genre is Blues, Meteor and the Girls whose music genre is Pop which rank the 2nd, 5th and 8th reletively on the list above.

In [11]:
%%sql
WITH
    employee_customer AS
    (
    SELECT e.first_name||" "||e.last_name employee_name,
           e.employee_id,
           e.hire_date hire_date,
           c.customer_id customer_id
    FROM employee e 
    INNER JOIN customer c ON e.employee_id=c.support_rep_id)
SELECT ec.employee_name,
       ec.hire_date,
       SUM(i.total) total_dollar_employee
FROM invoice i 
INNER JOIN employee_customer ec ON i.customer_id=ec.customer_id
GROUP BY 1
ORDER BY 2 desc;

 * sqlite:///chinook.db
Done.


[('Steve Johnson', '2017-10-17 00:00:00', 1393.920000000002),
 ('Margaret Park', '2017-05-03 00:00:00', 1584.0000000000034),
 ('Jane Peacock', '2017-04-01 00:00:00', 1731.5100000000039)]

While there is a 20% difference in sales between Jane (the top employee) and Steve (the bottom employee), the difference roughly corresponds with the differences in their hiring dates.

In [13]:
%%sql
WITH country_or_other AS
    (
     SELECT
        CASE
            WHEN (
                   SELECT COUNT(*)
                   FROM customer
                   WHERE country=c.country) = 1 THEN "Other"
            ELSE c.country
            END AS country,
        c.customer_id,
        il.*
      FROM invoice_line il
      INNER JOIN invoice i ON i.invoice_id=il.invoice_id
      INNER JOIN customer c ON c.customer_id=i.customer_id)
SELECT
    country,
    COUNT(DISTINCT customer_id) customers,
    SUM(unit_price) total_sales,
    SUM(unit_price) / COUNT(DISTINCT customer_id) customer_lifetime_value,
    SUM(unit_price) / COUNT(DISTINCT invoice_id) avarage_order
FROM (
    SELECT 
        co.*,
        CASE
            WHEN co.country='Other' THEN 1
            ELSE 0
        END AS sort
    FROM country_or_other co)
GROUP BY 1
ORDER BY sort ASC, 3 DESC;

 * sqlite:///chinook.db
Done.


[('USA', 13, 1040.490000000008, 80.03769230769292, 7.942671755725252),
 ('Canada', 8, 535.5900000000034, 66.94875000000043, 7.047236842105309),
 ('Brazil', 5, 427.68000000000245, 85.53600000000048, 7.011147540983647),
 ('France', 5, 389.0700000000021, 77.81400000000042, 7.781400000000042),
 ('Germany', 4, 334.6200000000016, 83.6550000000004, 8.161463414634186),
 ('Czech Republic', 2, 273.24000000000103, 136.62000000000052, 9.108000000000034),
 ('United Kingdom', 3, 245.52000000000078, 81.84000000000026, 8.768571428571457),
 ('Portugal', 2, 185.13000000000022, 92.56500000000011, 6.3837931034482835),
 ('India', 2, 183.1500000000002, 91.5750000000001, 8.72142857142858),
 ('Other', 15, 1094.9400000000085, 72.99600000000056, 7.448571428571486)]

Based on the data, there may be opportunity in the following countries:
* Czech Republic
* UK
* India

It's worth keeping in mind that because the amount of data from each of these countries is relatively low. Because of this, we should be cautious spending too much money on new marketing campaigns, as the sample size is not large enough to give us high confidence. A better approach would be to run small campaigns in these countries, collecting and analyzing the new customers to make sure that these trends hold with new customers.

In [14]:
%%sql
WITH invoice_d AS 
    (SELECT * FROM invoice_line il 
     LEFT JOIN track t ON t.track_id=il.track_id ),

    album_d AS 
    (SELECT COUNT(track_id) tracks_in_album, 
     album_id 
     FROM track 
     GROUP BY 2),
    
    final_table AS 
    (SELECT  ind.invoice_id,
             COUNT(ind.track_id) count_track_id_in_invoice,
             MIN(ind.track_id) rep_track_id,
             COUNT(distinct ind.album_id) albums_id_in_invoice_id,
             ind.album_id rep_album_id,
             ad.tracks_in_album 
     FROM invoice_d ind  
     LEFT JOIN album_d ad ON ad.album_id=ind.album_id 
     GROUP BY 1)

    
SELECT  
    COUNT(invoice_id) "number of invoices",
    ROUND((CAST(COUNT(invoice_id)AS FLOAT)/(SELECT COUNT(invoice_id) FROM final_table))*100,2) "invoice in %", 
    CASE WHEN
            (SELECT il.track_id 
             FROM invoice_line il 
             WHERE il.invoice_id=ft.invoice_id 
             EXCEPT 
             SELECT t.track_id FROM track t 
             WHERE t.album_id=ft.rep_album_id)IS NULL

         AND 

            (SELECT t.track_id FROM track t 
             WHERE t.album_id=ft.rep_album_id 
             EXCEPT 
             SELECT il.track_id 
             FROM invoice_line il 
             WHERE il.invoice_id=ft.invoice_id)IS NULL 

        THEN "ALBUM"

        ELSE "SINGLE" 

    END AS purchase_type

FROM final_table ft

GROUP BY purchase_type;

 * sqlite:///chinook.db
Done.


[(114, 18.57, 'ALBUM'), (500, 81.43, 'SINGLE')]

Album purchases account for 18.6% of purchases. Based on this data, I would recommend against purchasing only select tracks from albums from record companies, since there is potential to lose one fifth of revenue.

